FilterRules/degreesofseparation.py

```diff
@@ -0,0 +1,203 @@
+#
+# -------------------------------------------------------------------------
+class DegreesOption(MyInteger):
+    """Number option for filter editor."""
+
+    def __init__(self, database):
+        MyInteger.__init__(self, 1, 32)
+        self.set_tooltip_text(_("Number of degrees of separation from"
+                                " person."))
+
+
+# -------------------------------------------------------------------------
+#
+# IncPartners boolean option
+#
+# -------------------------------------------------------------------------
+class InclPartner(MyBoolean):
+    """Boolean option for filter editor."""
+
+    def __init__(self, database):
+        MyBoolean.__init__(self, _('Include Partners'))
+        self.set_tooltip_text(_("Include all partners."))
+        self.set_active(True)
+
+    def set_text(self, val):
+        """Set the checkbox active."""
+        is_active = bool(int(val))
+        self.set_active(is_active)
+
+
+# -------------------------------------------------------------------------
+#
+# IncPartners boolean option
+#
+# -------------------------------------------------------------------------
+class InclAllParents(MyBoolean):
+    """Boolean option for filter editor."""
+
+    def __init__(self, database):
+        MyBoolean.__init__(self, _('Include all parents'))
+        self.set_tooltip_text(_("Include parents with relationship foster,"
+                                " adopted, stepchild, etc."))
+        self.set_active(True)
+
+    def set_text(self, val):
+        """Set the checkbox active."""
+        is_active = bool(int(val))
+        self.set_active(is_active)
+
+
+# -------------------------------------------------------------------------
+#
+# Degrees of separation filter rule class
+#
+# -------------------------------------------------------------------------
+class DegreesOfSeparation(Rule):
+    """Filter rule that matches relatives by degrees of separation."""
+
+    labels = [_('ID:'),
+              (_("Degrees:"), DegreesOption),
+              ('', InclPartner),
+              ('', InclAllParents)]
+    name = _('People separated less than <N> degrees of <person>')
+    category = _("Relationship filters")
+    description = _("Filter rule that matches relatives by degrees of"
+                    " separation")
+
+    def prepare(self, db, user):
+        """Prepare a refernece list for the filter."""
+        self.db = db
+        self.persons = set()
+        self.ancestors = list()
+        root_handle = self.__get_root_handle()
+
+        self.__get_ancestors(root_handle)
+        for ancestor in self.ancestors:
+            self.__get_desc(ancestor, 0)
+
+        get_partners = bool(int(self.list[2]))
+        if get_partners:
+            self.__get_partners()
+
+    def __get_root_handle(self):
+        """Get the handle of the starting person."""
+        pid = self.list[0]
+        person = self.db.get_person_from_gramps_id(pid)
+        root_handle = person.get_handle()
```

Comment:
You need to make a test for no person here (None), in case the filter is in use with a different tree and root handle is not defined.  See https://gramps-project.org/bugs/view.php?id=12426